### ISSUE TITLE
Temporal control over Templates is needed

### ISSUE THREAD

**Issue by sbts** [ledgersmb/LedgerSMB#2477]
We need to have temporal control over templates.

If an invoice is printed today with one set of Terms and Conditions, business name etc, shouldn't it always print the same way.
Even if the template is changed in the future.

Additionally, if we are printing an invoice that has previously been printed, the button should be renamed to 
> rePrint

**Comment by ehuelsmann** [comment]
We should however allow for the use-case where someone is designing an invoice and needs to be able to repeatedly print the invoice for "template development purposes".

**Comment by einhverfr** [comment]
I don't think that is the right soltuion.  I think a better one would be to attach the printed file to the invoice.

**Comment by ehuelsmann** [comment]
That *would* solve another problem: you can't see what your customer saw when you mailed them, if you don't include yourself in the CC...

**Comment by sbts** [comment]
Attaching the file certainly is an option.
Although it's not a requirement to seeing what a customer saw. As long as the templates are temporally stored.
Oh. And the email body as well of course.

I think temporal templates (and other data) is probably a better use of space in the long run. We already have the data, we just need to ensure we can recreate any output as it was when originally generated

**Comment by einhverfr** [comment]
I don't know that this solves all the problems though.  Currently we do not store all data that goes into rendering the template in the invoice tables of the database.  So for example, there is no guarantee that part numbers don't change.

Email body is supposed to go into internal notes btw.

**Comment by ehuelsmann** [comment]
hmmm. but shouldn't partnumbers (generally: all parts data other than the inventory tracking) have temporal control too?

**Comment by sbts** [comment]
Recent versions of Postgres have temporal capabilities.
We should look into those and I'd suggest implementation against select columns and/or tables may solve many issues with data reproduction

**Comment by freelock** [comment]
Can this issue be closed with the new workflow in 1.9?

**Comment by ehuelsmann** [comment]
All that is missing is the actual attachment of the invoice when the print button is clicked...
